macro/src/upcasts.rs

```diff
@@ -0,0 +1,137 @@
+    substitution::{Substitute, Substitution},
+};
+
+/// A map storing the transitive upcasts for each class that we are generated (and potentially additional classes).
+///
+/// There is one caveat: we only compute the transitive superclasses based on the classes that are input to
+/// the proc macro. The problem is that we can only inspect the tokens presented to us. While we could reflect
+/// on the Java classes directly, we don't know what subset of the supertypes the user has chosen to reflect into
+/// Rust. Therefore, we stop our transitive upcasts at the "water's edge" -- i.e., at the point where we
+/// encounter classes that are outside our package.
+#[derive(Default, Debug)]
+pub struct Upcasts {
+    map: BTreeMap<DotId, ClassUpcasts>,
+}
+
+#[derive(Debug)]
+pub struct ClassUpcasts {
+    generics: Vec<Id>,
+    extends: BTreeSet<ClassRef>,
+}
+
+impl<'a> FromIterator<&'a ClassInfo> for Upcasts {
+    fn from_iter<T: IntoIterator<Item = &'a ClassInfo>>(iter: T) -> Self {
+        let mut upcasts = Upcasts::default();
+
+        for class_info in iter {
+            upcasts.insert_direct_upcasts(class_info);
+        }
+
+        upcasts.insert_hardcoded_upcasts();
+
+        upcasts.compute_transitive_upcasts();
+
+        upcasts
+    }
+}
+
+impl Upcasts {
+    /// Returns the transitive superclasses / interfaces of `name`.
+    /// These will reference generic parameters from in the class declaration of `name`.
+    pub fn upcasts_for_generated_class(&self, name: &DotId) -> &BTreeSet<ClassRef> {
+        &self.map[name].extends
+    }
+
+    /// Insert the direct (declared by user) superclasses of `class` into the map.
+    fn insert_direct_upcasts(&mut self, class: &ClassInfo) {
+        let mut upcasts = ClassUpcasts {
+            generics: class.generics.iter().map(|g| g.id.clone()).collect(),
+            extends: BTreeSet::default(),
+        };
+
+        upcasts.extends.insert(class.this_ref());
+
+        for c in class.extends.iter().chain(&class.implements) {
+            upcasts.extends.insert(c.clone());
+        }
+
+        upcasts.extends.insert(ClassRef {
+            name: DotId::object(),
+            generics: vec![],
+        });
+
+        let old_value = self.map.insert(class.name.clone(), upcasts);
+        assert!(old_value.is_none());
+    }
+
+    fn insert_hardcoded_upcasts(&mut self) {
+        let mut insert = |c: DotId, d: DotId| {
+            self.map
+                .entry(c)
+                .or_insert_with(|| ClassUpcasts {
+                    generics: vec![],
+                    extends: BTreeSet::default(),
+                })
+                .extends
+                .insert(ClassRef {
+                    name: d,
+                    generics: vec![],
+                });
+        };
+
+        insert(DotId::runtime_exception(), DotId::exception());
+        insert(DotId::exception(), DotId::throwable());
+        insert(DotId::throwable(), DotId::object());
```

Comment:
`insert_direct_upcasts` generates reflexive upcast relations, but the current `insert_hardcoded_upcasts` does not. I'm not sure if this is intended.

For example, if `compute_transitive_upcasts` starts with `{ E -> [T], RE -> [E], T -> [O] }` where `RE` stands for `RuntimeException` etc the current execution is:
* first loop iteration:
    * the inner loop iteration `n = E` ends with `{ E -> [T, O], RE -> [E], T -> [O] }`
    * the inner loop iteration`n = RE` ends with `{ E -> [T, O], RE -> [E, T, O], T -> [O] }`
    * the inner loop iteration`n = T` ends with `{ E -> [T, O], RE -> [E, T, O], T -> [O] }`
* second loop iteration:
    * nothing changes; the loop terminates.
* result: `E <: E`, `RE <: RE`, `T <: T` and `O <: O` are not in the map.

To make it consistent:
```suggestion
        insert(DotId::runtime_exception(), DotId::runtime_exception());
        insert(DotId::runtime_exception(), DotId::exception());
        insert(DotId::exception(), DotId::exception());
        insert(DotId::exception(), DotId::throwable());
        insert(DotId::throwable(), DotId::throwable());
        insert(DotId::throwable(), DotId::object());
        insert(DotId::object(), DotId::object());
```
Alternatively, the closure should also insert the reflexive relations for both arguments.

Reply:
It's intentional, but not harmful to include them. The reason is that the reflexive relations only matter when you are generating the `Exception` etc types, and in that case, you will get the reflexive relation from the `insert_direct_upcasts` call

Reply:
I couldn't bring myself to do the clones needed to include them.